src/odbc_connection.cpp

```diff
@@ -2123,43 +2126,73 @@ SQLRETURN ODBCConnection::FetchAll(QueryData *data) {
 
     // Iterate over each column, putting the data in the row object
     for (int i = 0; i < data->columnCount; i++) {
-      row[i].size = data->columns[i]->StrLen_or_IndPtr;
-      if (row[i].size == SQL_NULL_DATA || row[i].size == SQL_NO_TOTAL) {
-        row[i].char_data = NULL;
+
+      if (data->columns[i]->StrLen_or_IndPtr == SQL_NULL_DATA) {
+        row[i].size = SQL_NULL_DATA;
       } else {
         switch (data->columns[i]->bind_type) {
-          case SQL_C_DOUBLE:
-            row[i].double_data = *(SQLDOUBLE*)data->boundRow[i];
-            break;
+        case SQL_C_DOUBLE:
+          row[i].double_data = *(SQLDOUBLE*)data->boundRow[i];
+          break;
 
-          case SQL_C_UTINYINT:
-            row[i].tinyint_data = *(SQLCHAR*)data->boundRow[i];
-            break;
+        case SQL_C_UTINYINT:
+          row[i].tinyint_data = *(SQLCHAR*)data->boundRow[i];
+          break;
 
-          case SQL_C_USHORT:
-            row[i].smallint_data = *(SQLUSMALLINT*)data->boundRow[i];
-            break;
+        case SQL_C_USHORT:
+          row[i].smallint_data = *(SQLUSMALLINT*)data->boundRow[i];
+          break;
 
-          case SQL_C_SLONG:
-            row[i].integer_data = *(SQLINTEGER*)data->boundRow[i];
-            break;
+        case SQL_C_SLONG:
+          row[i].integer_data = *(SQLINTEGER*)data->boundRow[i];
+          break;
 
-          case SQL_C_UBIGINT:
-            row[i].ubigint_data = *(SQLUBIGINT*)data->boundRow[i];
-            break;
+        case SQL_C_UBIGINT:
+          row[i].ubigint_data = *(SQLUBIGINT*)data->boundRow[i];
+          break;
 
-          case SQL_C_WCHAR:
-            row[i].wchar_data = new SQLWCHAR[row[i].size + 2]();
-            memcpy(row[i].wchar_data, data->boundRow[i], row[i].size);
-            break;
+        case SQL_C_BINARY:
+          row[i].char_data = new SQLCHAR[row[i].size]();
+          memcpy(row[i].char_data, data->boundRow[i], row[i].size);
+          break;
 
-          case SQL_C_CHAR:
-          default:
-            row[i].char_data = new SQLCHAR[row[i].size + 2]();
-            memcpy(row[i].char_data, data->boundRow[i], row[i].size);
-            break;
-        }
-        row->bind_type = data->columns[i]->bind_type;
+        case SQL_C_WCHAR:
+          row[i].size = wcslen((const wchar_t *)data->boundRow[i]);
```

Comment:
This will only work on Windows. On Linux, macOS, PASE, and other non-Windows platforms a wchar_t will be 4 bytes, while the SQLWCHAR is 2 bytes and thus the only time wcslen will find the 4 byte null terminator is by accident: when the two bytes following the buffer _happen_ to be both 0.

Reply:
Created a custom function to count the number of 2-byte code units:
```
size_t strlen16(const char16_t* string)
{
   if (!string) return -1;
   const char16_t* str = string;
   while(*str) str++;
   return str - string;
}
```

Reply:
Should probably return 0 or `NULL`.

Reply:
Or should it just not check, spin the roulette wheel, and result in undefined behavior like `strlen`?

Reply:
I would go for whatever `strlen` does.